### PR TITLE
LDAP: attempt to connect to backup server again, if main server is no…

### LINKED ISSUE #18701: Backup Host for LDAP server does not automatically connect
### Steps to reproduce
1. Setup an LDAP connection with a Backup server
2. Block the network access to the main LDAP server -- To simulate failure, I've added an entry to the owncloud host file pointing the ldap server to localhost, when this is in place users are unable to login to the web interface with the error "An internal error occurred.
3.

### Expected behaviour
Once the Main LDAP server is unreachable, the Backup server should be used for connections

### Actual behaviour
Errors are displayed when a user attempts to login. Logins will happen if the check box disable main server is enabled, so we are certain the backup host configuration is correct.
```
{"app":"handleLogin","message":"OC\\ServerNotAvailableException: Lost connection to LDAP server.","level":4,"time":"2015-08-28T09:23:44-06:00"}
{"app":"handleLogin","message":"OC\\ServerNotAvailableException: Lost connection to LDAP server.","level":4,"time":"2015-08-28T09:25:31-06:00"} 
```

### Server configuration
**Operating system**: RHEL 6

**Web server:** Apache

**Database:** Oracle

**PHP version:** 5.4.40

**ownCloud version:** 7.0.8

**Updated from an older ownCloud or fresh install:** Updated

**List of activated apps:** See config report https://s3.owncloud.com/owncloud/index.php/apps/files?dir=%2FShared%2Fowncloud%2Fsupport%2Fgithub-issues%2Fcore%2F18701

**The content of config/config.php:**

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** NO

**Are you using an external user-backend, if yes which one:** LDAP

#### LDAP configuration
See config report

#### Web server error log
```
Insert your webserver log here
```

#### ownCloud log (data/owncloud.log)
```
{"app":"handleLogin","message":"OC\\ServerNotAvailableException: Lost connection to LDAP server.","level":4,"time":"2015-08-28T09:23:44-06:00"}
{"app":"handleLogin","message":"OC\\ServerNotAvailableException: Lost connection to LDAP server.","level":4,"time":"2015-08-28T09:25:31-06:00"} 
```

### PR DESCRIPTION
**Pull request by blizzz**
…t available. Fixes #18701 

First things first, we are now catching an offline server when initially connecting to it and try to bind to the backup server. However, if we are connected to the first one and sometime in the runtime the connection gets lost, we die with an exception instead of retrying the connection. Basically this is what was done back then when the fallback was working (only that we did not throw any exception in that cases and caused weird side effects).

How to test:
1. Have an non-working LDAP Host
2. Provide a working LDAP Backup Server
(I put the host and port info of my test LDAP into the backup host and port fields and changed the main ones to something not exisiting)
3. Ensure that configuration is active
4. Go to Users page
Before: Exception
Now: Successfully connects to the backup

### CODE REVIEW

**Comment by blizzz** [comment]
Pleaser test and review @johnripper1 @pierrejochem @MorrisJobke @patrick1990 @davitol 

@karlitschek request backport confirmation down to 8.0

File: apps/user_ldap/tests/integration/lib/integrationtestbackupserver.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+			'ldap_backup_host' => $originalHost,
+			'ldap_backup_port' => $originalPort,
+		]);
+	}
+
+	/**
+	 * tests that a backup connection is being used when the main  LDAP server
+	 * is offline
+	 *
+	 * Beware: after starting docker, the LDAP host mind not be ready yet, thus
```
Comment (phil-davis) [inline apps/user_ldap/tests/integration/lib/integrationtestbackupserver.php]: s/mind/might/

**Comment by MorrisJobke** [comment]
I tested this. Before the user list was empty. With this patch it is visible. Also no errors in the logs :+1:

**Comment by MorrisJobke** [comment]
Also login works now properly ;)

File: apps/user_ldap/tests/integration/lib/integrationtestbackupserver.php (hunk 1 above)
Comment (blizzz) [inline apps/user_ldap/tests/integration/lib/integrationtestbackupserver.php]: ty

**Comment by blizzz** [comment]
Also here, one more reviewer needed, cc @PVince81

**Comment by PVince81** [comment]
:+1: code makes sense.

@karlitschek backport ? (original ticket was for 8.2.1)

**Comment by karlitschek** [comment]
please backport :+1:

**Comment by PVince81** [comment]
@karlitschek I just noticed that @blizzz already asked above, backport down to 8.0.

@blizzz can you make the PRs ? Thanks :smile:

**Comment by davitol** [comment]
:+1: Tested and WFM

**Comment by blizzz** [comment]
Backports, heiß und fettig:
8.2: #20357 
8.1: #20362 
8.0: #20363

**Pull request merged**